### ISSUE TITLE
'Text (FreeType 2)' element only reading first line of file (when 'Read from file' is checked)

### ISSUE THREAD

**Issue by raleighlittles** [obsproject/obs-studio#7595]
### Operating System Info

Ubuntu 22.04

### OBS Studio Version

28.0.3

### OBS Studio Log URL

https://obsproject.com/logs/VJ5xkLXGk9ICAlaP

### Expected Behavior

Expected behavior: Each line of the text file would be read, and a multi-line string would be shown.

### Current Behavior

Only the first line of the text file is read, and other lines aren't shown. See screenshot: 

![screenshot](https://i.imgur.com/oFbrTHL.png)

To summarize:

* I have 'Text(FreeType 2)' set to read a text file. The text file has 4 lines. Only the first line of the file is being rendered.

### Steps to Reproduce

1. Create a text file with multiple lines of text.
2. Create a 'Text (FreeType 2)' source element
3. Check "Read from file", and in the 'Text File' field, select your multi-line file.
4. Observe that only one line of output is shown, regardless of how many lines your actual file has.
...

**Comment by kkartaltepe** [comment]
I cannot replicate, provide a sample file.

**Comment by raleighlittles** [comment]
Ah okay I figured out what the actual problem is. When a file is read, OBS is incorrectly treating the null character in a file as the end-of-file indicator.

I ran `cat` with the extended option and saw this:

```
➜ cat -A test.txt 
Manufacturer: EVGA^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
Product: EVGA XR1 Lite Capture Box Video^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
Serial number: 385203091463^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
(0x3842 : 0x310a)%   
```
If I remove the null characters, the issue goes away.

----

I bet the issue is somewhere here: https://github.com/obsproject/obs-studio/blob/1af1a432e2edc29f5b914350cdfdd44fbe480186/plugins/text-freetype2/text-functionality.c#L421

@kkartaltepe

**Comment by raleighlittles** [comment]
[test.txt](https://github.com/obsproject/obs-studio/files/9793734/test.txt)

**Comment by PranavJoshi18** [comment]
Hey! Maybe I can help with this issue if it is still open.... I know the basic code for reading multiple lines in C. I may help with the related issue.. Actually it would be my first time contributing so please don't mind if I ask a lot of question 😅

**Comment by lourencop01** [comment]
Hey, has this issue been solved? I have just tested the test.txt file provided above and it just read the first line. But I'm guessing the problem comes from the txt file and not OBS itself.

If it is not solved, I'd like to work on it! :)

**Comment by kkartaltepe** [comment]
Yes this still occurs, we hoped to allow some quirks in utf8 to make a simple fix, but also someone has previously submitted the change with proper length handling in #7622 that would need to be resubmitted and reviewed.

**Comment by lourencop01** [comment]
Ok, so from what I've read on #7622 and #7628 you guys are looking for a solution that represents the null bytes as invalid character glyphs.

From your message I understand that the change in #7622 did that, correct? You just needed it to be resubmitted and reviewed.

If I could help in any way related to this let me know as I am currently looking for a bug to fix here on OBS! :)

Or just mention any other issue I could get my hands on (I'm running macOS Sonoma 14.3.1)

**Comment by kkartaltepe** [comment]
> Ok, so from what I've read on https://github.com/obsproject/obs-studio/pull/7622 and https://github.com/obsproject/obs-studio/pull/7628 you guys are looking for a solution that represents the null bytes as invalid character glyphs.

No, we need someone to fix the string handling in this plugin and util to correctly pass/use the length and fix any other issues related to strings containing the null byte.

> From your message I understand that the change in https://github.com/obsproject/obs-studio/pull/7622 did that, correct? You just needed it to be resubmitted and reviewed.

Appeared to do this, but hasnt been reviewed. Someone would need to do something similar to that PR. It wasnt reviewed so who knows if its correct or broken or what.

**Comment by lourencop01** [comment]
Alright! If that's the case I'd love to take on this issue :)

**Comment by Anurag-Modak** [comment]
Is there anyone working on this issue?

**Comment by Anurag-Modak** [comment]
> Ah okay I figured out what the actual problem is. When a file is read, OBS is incorrectly treating the null character in a file as the end-of-file indicator.
> 
> I ran `cat` with the extended option and saw this:
> 
> ```
> ➜ cat -A test.txt 
> Manufacturer: EVGA^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
> Product: EVGA XR1 Lite Capture Box Video^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
> Serial number: 385203091463^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@$
> (0x3842 : 0x310a)%   
> ```
> 
> If I remove the null characters, the issue goes away.
> 
> I bet the issue is somewhere here:
> 
> https://github.com/obsproject/obs-studio/blob/1af1a432e2edc29f5b914350cdfdd44fbe480186/plugins/text-freetype2/text-functionality.c#L421
> 
> @kkartaltepe

we should  use the file size directly in the wide character conversion function to avoid issues with null characters in the file content.

**Comment by RickeyEarl** [comment]
I'm looking into this. Will update end of week.

**Comment by kkartaltepe** [comment]
Thanks for the interest in the issue, but I finally got some time to fix up the old PR that I had pointed folks towards so we will probably close this out with that PR. If you get a chance to test and notice anything not working with the pending PR please leave a comment.
